src/tqec/circuit/detector/detector_util.py

```diff
@@ -0,0 +1,406 @@
+    seen_measurement_at_tail: bool = False
+
+    def clear(self) -> None:
+        self.cur_fragment.clear()
+        self.end_stabilizer_sources.clear()
+        self.sources_for_next_fragment.clear()
+        self.begin_stabilizer_sources.clear()
+        self.seen_reset_at_head = False
+        self.seen_measurement_at_tail = False
+
+    def to_fragment(self) -> Fragment:
+        return Fragment(
+            circuit=self.cur_fragment.copy(),
+            end_stabilizer_sources=list(self.end_stabilizer_sources),
+            begin_stabilizer_sources=list(self.begin_stabilizer_sources),
+            sources_for_next_fragment=list(self.sources_for_next_fragment),
+        )
+
+
+def split_stim_circuit_into_fragments(
+    circuit: stim.Circuit,
+) -> list[Fragment | FragmentLoop]:
+    """Split the circuit into fragments."""
+    fragments = []
+    state = SplitState()
+    for moment in iter_stim_circuit_by_moments(circuit):
+        if isinstance(moment, stim.CircuitRepeatBlock):
+            if state.cur_fragment:
+                fragments.append(state.to_fragment())
+                state.clear()
+            body_fragments = split_stim_circuit_into_fragments(moment.body_copy())
+            fragments.append(
+                FragmentLoop(fragments=body_fragments, repetitions=moment.repeat_count)
+            )
+        elif has_measurement(moment, check_are_all_measurements=True):
+            state.cur_fragment += moment
+            state.begin_stabilizer_sources.extend(
+                collapse_pauli_strings_at_moment(moment, is_reset=False)
+            )
+            state.seen_measurement_at_tail = True
+            state.sources_for_next_fragment.extend(
+                collapse_pauli_strings_at_moment(moment, is_reset=True)
+                if has_reset(moment, check_are_all_resets=False)
+                else []
+            )
+        elif has_reset(moment, check_are_all_resets=True):
+            pauli_strings = collapse_pauli_strings_at_moment(moment, is_reset=True)
+            if state.seen_measurement_at_tail:
+                state.cur_fragment += moment
+                state.sources_for_next_fragment.extend(pauli_strings)
+            elif state.seen_reset_at_head:
+                state.cur_fragment += moment
+                state.end_stabilizer_sources.extend(pauli_strings)
+            else:
+                if state.cur_fragment:
+                    fragments.append(state.to_fragment())
+                    state.clear()
+                state.cur_fragment += moment
+                state.end_stabilizer_sources.extend(pauli_strings)
+                state.seen_reset_at_head = True
+
+        elif state.seen_measurement_at_tail:
+            fragments.append(state.to_fragment())
+            state.clear()
+            state.cur_fragment += moment
+        else:
+            state.cur_fragment += moment
+    if state.cur_fragment:
+        fragments.append(state.to_fragment())
+    return fragments
```

Comment:
This function is quite complex to understand. I think it would help to provide in comments a few invariants that are always checked through the for loop. A few examples of invariants that might be useful to help someone understanding that function:

- `state` stores and gather the necessary information to build a `Fragment` and is reset each time the final `Fragment` is constructed with `SplitState.to_fragment`.
- `state` represents a complete `Fragment` when any of the following conditions is met: `[list here]`